source/testcommon/README.md

```diff
@@ -1,9 +1,153 @@
-# Test Common Py
+# Testcommon utility tool
 
-This is common testing utilities for Python projects.
+This python packages includes tools and utility function for making testing easier.
+
+## Usage
+
+### Scenario testing
+
+To use testcommon for the scenario testing we need to set up three things.
+
+#### 1) Use the standard folder structure setup
+
+```plaintext
+├── scenario_name/
+    ├── when/
+        ├── input1.csv
+        └── ...
+    └── then/
+        ├── output1.csv
+        └── ...
+    └── test_output.py
+└── conftest.py
+```
+
+Input files are stored in the `/when` subfolder, while the expected results are stored in the `/then` subfolder.
+We prefix scenario names with `given_...` to enhance readability. As if we ready it line-by-line it makes sense.
+This is best shown with an example.
+
+```plaintext
+├──given_energy_scenario
+    ├── when/
+        ├── energy.csv
+        ├── energy_per_brp.csv
+        └── energy_per_es.csv
+    └── then/
+        └── energy_v1.csv
+```
+
+Notice, that there is not a one-to-one relation between the `/then` and `/when` files.
+This is because we may have to perform various transformations, such as aggregating or joining tables, to achieve the
+desired
+output. (i.e., `/then` file).
+
+#### 2) conftest.py
+
+The `conftest.py` file is a special pytest file where `fixtures` can be defined. `Fixtures` are small function that will
+run
+before the actual test runs (i.e. the `test_output.py`). As such we can use it for setting up the testing environment
+and
+therefore make the actual tests small and concise (i.e. limiting code duplication).
+
+Each `fixture` will have a `scope` defining how long the fixture is active and when it is set up and torn down.
+There are different types:
+
+1. Function: the default scope, the fixture is destroyed at the end of the test.
+1. Class: the fixture is destroyed during teardown of the last test in the class.
+1. Module: the fixture is destroyed during teardown of the last test in the module.
+1. Package: the fixture is destroyed during teardown of the last test in the package.
+1. Session: the fixture is destroyed at the end of the test session.
```

Comment:
Rather than explaining scope we should document why we use "module" and why it's important.

Iit ensures that (1) fixture is reused for each then-file in the scenario, and (2) that different tests doesn't use the same data. (1) is important when the fixture is slow - e.g. because of a long-running transformation. In fact this is also the main reason for the whole fixture design/idea. Which is important as it creates a slightly odd test setup where the "act" part is actually executed before the (pytest) test itself is executed. Feel free to reach out if needed.